tools/scripts/db.py

```diff
@@ -0,0 +1,275 @@
+        # commit
+        # cur.execute(SQL("""
+        #     INSERT INTO {}( {} )
+        #     SELECT {} FROM {} WHERE state = 'COMPLETED'
+        # """).format(Identifier(table), SQL(cols), cols, Identifier(table)))
+
+        print("added rows:", cur.rowcount)
+        cur.execute("COMMIT")
+
+
+def copy_trial(trial_id: int) -> None:
+    """
+    copy a single trial with associated metrics using CTE.
+    """
+    trial_cols = get_table_col_names("trials") - {"id"}
+    trial_cols_str = ", ".join(trial_cols)
+    steps_cols = get_table_col_names("raw_steps") - {"id", "trial_id"}
+    steps_cols_str = ", ".join(steps_cols)
+    prefixed_steps_cols = ", ".join([f"rs.{col}" for col in steps_cols])
+    validations_cols = get_table_col_names("raw_validations") - {"id", "trial_id"}
+    validations_cols_str = ", ".join(validations_cols)
+    prefixed_validations_cols = ", ".join([f"rv.{col}" for col in validations_cols])
+
+    with db_cursor() as cur:
+        query = f"""
+WITH replicated_trials AS (
+INSERT INTO trials ({trial_cols_str})
+SELECT {trial_cols_str}
+FROM trials
+WHERE id = %s
+RETURNING id
+), replicated_steps AS (
+INSERT INTO raw_steps (trial_id, {steps_cols_str})
+SELECT rt.id, {prefixed_steps_cols}
+FROM replicated_trials rt
+JOIN raw_steps rs ON rs.trial_id = %s
+RETURNING trial_id, id AS new_step_id
+)
+INSERT INTO raw_validations (trial_id, {validations_cols_str})
+SELECT rt.id, {prefixed_validations_cols}
+FROM replicated_trials rt
+JOIN raw_validations rv ON rv.trial_id = %s;
+        """
+        cur.execute(query, (trial_id, trial_id, trial_id))  # type: ignore
+        cur.execute("COMMIT")
+
+
+def copy_trials(multiplier=1, suffix="") -> dict:
+    table = "trials"
+    assert multiplier == 1  # disable this for now
+    cols = get_table_col_names(table) - {"id"}
+    cols_str = ", ".join(cols)
+
+    with db_cursor() as cur:
+        query = f"""
+-- modify the target table to add a new col called og_id
+ALTER TABLE {table} ADD COLUMN og_id int;
+"""
+        cur.execute(query)  # type: ignore
+
+        query = f"""
+-- insert the replicated rows populating the og_id column with the original id
+INSERT INTO {table}( {cols_str}, og_id )
+SELECT {cols_str}, id
+FROM {table}
+-- CROSS JOIN generate_series(1, {multiplier}) AS g
+{suffix};
+"""
+        cur.execute(query)  # type: ignore
+        replicated_trials = cur.rowcount
+
+        steps_cols = get_table_col_names("raw_steps") - {"id"} - {"trial_id"}
+        steps_cols_str = ", ".join(steps_cols)
+        prefixed_steps_cols = ", ".join([f"rs.{col}" for col in steps_cols])
+        # replicate raw_steps and update trial_id
+        query = f"""
+
+-- replicate raw_steps and keep the new step ids
+INSERT INTO raw_steps( {steps_cols_str}, trial_id )
+SELECT {prefixed_steps_cols}, trials.id
+FROM raw_steps rs
+INNER JOIN trials ON trials.og_id = rs.trial_id   -- also limits the steps to target trials. CHECK: which join type?
```

Comment:
currently the most expensive part.

Reply:
not sure why this is taking the most of amount time. maybe because of the join?

Reply:
I updated this to use a temp table to store the id mappings between the old and new trials and that improved this a lot as the joins are much less expensive.